[Bug] 既存ユーザーにGoogle OAuth連携を追加した際、email_verifiedが更新されない

## バグ概要
既存ユーザー（通常のメール・パスワード登録）に対してGoogle OAuth連携を追加する際、Googleが既にメール認証済みであるにも関わらず、`email_verified`フィールドが`false`のまま更新されない問題が発生しています。

### 期待される動作
1. 通常のユーザー登録（メール・パスワード）でユーザー作成（`email_verified: false`）
2. メール認証を完了せずに放置
3. 同じメールアドレスでGoogle OAuth認証でログイン
4. **期待**: 既存ユーザーにGoogle OAuth連携を追加し、Googleが既にメール認証済みなので`email_verified`を`true`に更新する
5. ログイン後、すぐにアプリを利用できる

### 実際の動作
1. 通常のユーザー登録（メール・パスワード）でユーザー作成（`email_verified: false`）
2. メール認証を完了せずに放置
3. 同じメールアドレスでGoogle OAuth認証でログイン
4. **実際**: 既存ユーザーにGoogle OAuth連携を追加するが、`email_verified`は`false`のまま
5. ログイン後、403 Forbiddenエラーが発生し、アプリを利用できない

## 再現手順

1. 通常のユーザー登録（メール・パスワード）を行う
2. メール認証を完了せずに放置（`email_verified: false`のまま）
3. 同じメールアドレスでGoogle OAuth認証でログインを試みる
4. ログイン後、通知API等にアクセスする
5. 403 Forbiddenエラーが発生する

## 環境情報
- [x] **ブラウザ**: Chrome / Firefox / Safari / Edge / その他: すべて
- [x] **デバイス**: PC / タブレット / スマートフォン - すべて
- [x] **画面サイズ**: デスクトップ / タブレット / モバイル - すべて
- [x] **認証状態**: ログイン済み

## 発生頻度・影響度
- [x] **発生頻度**: 必ず発生（特定の条件下で100%再現）
- [x] **影響度**: 中（一部機能に影響）- 特定の順序でユーザー登録を行ったユーザーのみ

## エラー情報

### フロントエンドエラー
```javascript
GET https://api.ouchi-no-hatake.com/api/v1/notifications/unread_count 403 (Forbidden)
[ERROR] Failed to fetch unread count {error: 'Failed to fetch unread count'}
```

### バックエンドエラー
```
[5ba25c0e-804a-4078-a456-f9550081bfc2] Processing by Api::V1::NotificationsController#unread_count as */*
[5ba25c0e-804a-4078-a456-f9550081bfc2] {"timestamp":"2025-11-02T18:00:39Z","message":"Auth event: Authentication successful","user_id":9,"success":true}
[5ba25c0e-804a-4078-a456-f9550081bfc2] Filter chain halted as :check_email_verification rendered or redirected
[5ba25c0e-804a-4078-a456-f9550081bfc2] Completed 403 Forbidden in 6ms
```

## スクリーンショット・録画
（なし）

## 修正方針

### 開発方針チェック
> 修正時も開発思想「品質・保守性ファースト」「セキュリティファースト」「標準化」に従ってください

- [x] **品質・保守性**: 根本原因を解決する（対症療法ではなく）
- [x] **品質・保守性**: 同様の問題を防ぐ設計改善を検討する
- [x] **セキュリティ**: 修正により新たなセキュリティリスクを生まない
- [x] **標準化**: コーディング規約・パターンに従って修正する

### 修正対象ファイル
`backend/app/services/oauth_service.rb`

### 修正内容
既存ユーザーにOAuth連携を追加する際、Googleが既にメール認証済み（`auth_info[:email_verified] == true`）の場合、ユーザーの`email_verified`も`true`に更新する処理を追加する。

## 達成条件（Definition of Done）

### 修正要件
- [ ] 既存ユーザーにGoogle OAuth連携を追加する際、`email_verified`が`true`に更新される
- [ ] 新規ユーザー作成時の動作に影響を与えていない（`email_verified`が正しく設定される）
- [ ] 既にOAuth連携済みのユーザーの動作に影響を与えていない
- [ ] 既にメール認証済みのユーザーに対して無駄な更新を行わない
- [ ] RSpecテストを追加・更新し、すべてのテストが成功する
- [ ] 手動テストでの動作確認完了
  - 通常登録 → メール認証未完了 → Google OAuth認証 → 403エラーが出ないこと
  - 新規でGoogle OAuth認証 → 正常にログインできること
  - 既存のOAuth連携ユーザー → 正常にログインできること

## 参考資料
- Google OAuth認証仕様: Googleが既にメールアドレスを検証済み（`email_verified: true`）の場合、アプリ側で再度メール認証を要求する必要はない
- 関連PR: #317（Google OAuth認証機能実装）
- 影響を受けたユーザー: User ID 9（手動で`email_verified: true`に更新済み）

## 補足・注意事項

### 影響範囲
以下の条件をすべて満たすユーザーのみ影響を受ける：
1. 通常のユーザー登録（メール・パスワード）を先にした
2. メール認証を完了していない（`email_verified: false`）
3. その後、同じメールアドレスでGoogle OAuth認証でログインした

### セキュリティ考慮事項
- Google OAuth認証では、Google側で既にメールアドレスを検証済み
- `auth_info[:email_verified]`が`true`の場合のみ更新することで、不正な認証を防ぐ
- ログに記録することで、トラブルシューティングを容易にする

### テスト方針
`oauth_service_spec.rb`にテストケースを追加：
- 既存ユーザー（`email_verified: false`）にOAuth連携を追加した際、`email_verified`が`true`に更新されること
- 既存ユーザー（`email_verified: true`）にOAuth連携を追加した際、無駄な更新が行われないこと
- 新規ユーザー作成時の動作が変わらないこと

## 参考資料・補足
- 本バグは、Google OAuth認証機能実装（PR #317）の後に発見されました
- 緊急対応として、本番環境で手動修正を実施済み（User ID 9の`email_verified`を`true`に更新）
- 根本的な修正により、将来の同様の問題を防ぎます

--- a/backend/app/services/oauth_service.rb
+++ b/backend/app/services/oauth_service.rb
@@ -70,6 +70,13 @@ class OauthService < ApplicationService
         uid: auth_info[:uid],
         email: auth_info[:email]
       )
+
+      # Googleがメールアドレスをすでに検証済みの場合、ユーザーのemail_verifiedをtrueに更新
+      if auth_info[:email_verified] && !user.email_verified?
+        user.update!(email_verified: true)
+        Rails.logger.info "Updated email_verified to true for user #{user.id} via OAuth"
+      end
+
       return user
     end
 

--- a/backend/spec/services/oauth_service_spec.rb
+++ b/backend/spec/services/oauth_service_spec.rb
@@ -1,0 +1,164 @@
+require 'rails_helper'
+
+RSpec.describe OauthService, type: :service do
+  describe '.authenticate_google' do
+    let(:valid_credential) { 'valid_google_credential' }
+    let(:auth_info) do
+      {
+        provider: 'google',
+        uid: '1234567890',
+        email: 'test@example.com',
+        name: 'Test User',
+        email_verified: true
+      }
+    end
+
+    before do
+      # decode_google_tokenメソッドをスタブしてHTTP通信をモック
+      allow(OauthService).to receive(:decode_google_token)
+        .with(valid_credential)
+        .and_return(auth_info)
+    end
+
+    context '新規ユーザーの場合' do
+      it 'ユーザーが作成され、email_verifiedがtrueになる' do
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.to change(User, :count).by(1)
+
+        user = User.last
+        expect(user.email).to eq('test@example.com')
+        expect(user.name).to eq('Test User')
+        expect(user.email_verified).to be true
+        expect(user.password_digest).to be_nil
+      end
+
+      it 'UserProviderが作成される' do
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.to change(UserProvider, :count).by(1)
+
+        provider = UserProvider.last
+        expect(provider.provider).to eq('google')
+        expect(provider.uid).to eq('1234567890')
+        expect(provider.email).to eq('test@example.com')
+      end
+
+      it 'JWTトークンが返される' do
+        result = OauthService.authenticate_google(valid_credential)
+        expect(result[:token]).to be_present
+        expect(result[:user]).to be_a(User)
+      end
+    end
+
+    context '既存ユーザー（email_verified: false）にOAuth連携を追加する場合' do
+      let!(:existing_user) do
+        create(:user, email: 'test@example.com', email_verified: false)
+      end
+
+      it 'ユーザーが新規作成されない' do
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.not_to change(User, :count)
+      end
+
+      it 'UserProviderが作成される' do
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.to change(UserProvider, :count).by(1)
+
+        provider = UserProvider.last
+        expect(provider.user_id).to eq(existing_user.id)
+        expect(provider.provider).to eq('google')
+        expect(provider.uid).to eq('1234567890')
+      end
+
+      it 'email_verifiedがtrueに更新される' do
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.to change { existing_user.reload.email_verified }.from(false).to(true)
+      end
+
+      it '既存ユーザーが返される' do
+        result = OauthService.authenticate_google(valid_credential)
+        expect(result[:user].id).to eq(existing_user.id)
+      end
+    end
+
+    context '既存ユーザー（email_verified: true）にOAuth連携を追加する場合' do
+      let!(:existing_user) do
+        create(:user, :verified, email: 'test@example.com')
+      end
+
+      it 'email_verifiedはtrueのまま（無駄な更新なし）' do
+        expect(existing_user.email_verified).to be true
+        OauthService.authenticate_google(valid_credential)
+        expect(existing_user.reload.email_verified).to be true
+      end
+
+      it 'UserProviderが作成される' do
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.to change(UserProvider, :count).by(1)
+      end
+    end
+
+    context '既にOAuth連携済みのユーザーの場合' do
+      let!(:existing_user) do
+        create(:user, email: 'test@example.com', email_verified: true)
+      end
+      let!(:existing_provider) do
+        create(:user_provider,
+               user: existing_user,
+               provider: 'google',
+               uid: '1234567890',
+               email: 'test@example.com')
+      end
+
+      it 'ユーザーもUserProviderも新規作成されない' do
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.not_to change(User, :count)
+
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.not_to change(UserProvider, :count)
+      end
+
+      it '既存ユーザーが返される' do
+        result = OauthService.authenticate_google(valid_credential)
+        expect(result[:user].id).to eq(existing_user.id)
+      end
+    end
+
+    context 'Google Token検証が失敗した場合' do
+      before do
+        # decode_google_tokenがAuthenticationErrorを発生させるようにスタブ
+        allow(OauthService).to receive(:decode_google_token)
+          .with(valid_credential)
+          .and_raise(OauthService::AuthenticationError, 'OAuth authentication failed: Invalid token')
+      end
+
+      it 'AuthenticationErrorが発生する' do
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.to raise_error(OauthService::AuthenticationError, /Invalid token/)
+      end
+    end
+
+    context 'Client IDが一致しない場合' do
+      before do
+        # decode_google_tokenがAuthenticationErrorを発生させるようにスタブ
+        allow(OauthService).to receive(:decode_google_token)
+          .with(valid_credential)
+          .and_raise(OauthService::AuthenticationError, 'OAuth authentication failed: Invalid client ID')
+      end
+
+      it 'AuthenticationErrorが発生する' do
+        expect {
+          OauthService.authenticate_google(valid_credential)
+        }.to raise_error(OauthService::AuthenticationError, /Invalid client ID/)
+      end
+    end
+  end
+end

